[calebchongsj] iP

calebchongsj opened the pull request:
### **QuirkBot: A Java Chatbot Application**

QuirkBot is a **Java Chatbot application** designed to help users manage their tasks efficiently. With a focus on simplicity and efficiency, QuirkBot is your task management buddy. 

> "The best way to predict the future is to create it." – Peter Drucker

QuirkBot monitors all of your daily tasks without you losing track of them.
- **Task Management**: Easily add and remove tasks.
- **Quick and Efficient**: SUPER FAST responses for task updates.
- **Text-Based Interface**: Simple and intuitive user-friendly chat experience.
- **Free of Charge**: Enjoy all features without any cost.

**Getting Started:**
1. Download the application from [here](https://github.com/calebchongsj/ip/releases)📥
2. Double-click the downloaded file to run it.
3. Add your tasks via the chatbot.
4. Let QuirkBot manage your tasks for you 😊

**Key Features:**
- **Managing tasks**
- Managing deadlines _(coming soon)_
- Reminders _(coming soon)_

**Example Java Code:**
Here’s a snippet of the Java code used for adding a task:
```
// Adds a new task to the task list
public void addTask(String taskDescription) {
    Task currentTask = Parser.parseCommand(taskDescription);
    if (currentTask != null) {
        taskList.addTask(currentTask);
        screen.showTaskAdded(currentTask, taskList.size());
    }
}
```
**Agenda:**
- [ ] Review application features.
- [ ] Test new functionalities.
- [ ] Provide feedback.

**QuirkBot** is designed to be _straightforward_ and user-friendly, providing a seamless task management experience. If you have any questions or suggestions, leave them in the comments down below.

Thank you for reviewing this pull request! 😄

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

calebchongsj committed:
Level-0 increment. Renamed Duke.java to YourHelperBuddy.java and implemented a simple chatbot which greets the user and exits.

calebchongsj committed:
Level-1 increment. Implemented my chatbot to echo whatever command the user inputs until he enters "bye".

calebchongsj committed:
Level-2 increment. Added the ability to store whatever text entered by the user and display them back to the user when requested.

calebchongsj committed:
Level-3 increment. Add the ability to mark tasks as done or undone. Updated the status of the tasks accordingly.

calebchongsj committed:
Level-4 increment. Add support for tracking three types of tasks:

ToDos: tasks without any date/time attached to it
Deadlines: tasks that need to be done before a specific date/time
Events: tasks that start at a specific date/time and ends at a specific date/time

calebchongsj committed:
Level-5 increment. Modified the chatbot to deal with errors such as incorrect inputs entered by the user.

calebchongsj committed:
A-TextUiTesting increment. Conducted the semi-automated testing.

calebchongsj committed:
Level-6 increment. The chatbot supports the delete task functionality.

calebchongsj committed:
Level-7 increment. Save the tasks in a hard disk automatically and load the tasks when the chatbot starts up.

calebchongsj committed:
Merge branch 'branch-Level-7'

calebchongsj committed:
Level-8 increment. The chatbot understands dates and times for deadline and event tasks.

calebchongsj committed:
Level-8 increment. The chatbot understands dates and times for deadline and event tasks.

calebchongsj committed:
Level-9 increment. The chatbot understands dates and times for deadline and event tasks.

calebchongsj committed:
Merge branch 'branch-Level-8'

calebchongsj committed:
Level-8 increment. The chatbot understands dates and times for deadline and event tasks.

This reverts commit a4d05b33cb83f7226ea8f032b37eda8cf222447d.

calebchongsj committed:
Level-8 increment. The chatbot understands dates and times for deadline and event tasks.

This reverts commit a4d05b33cb83f7226ea8f032b37eda8cf222447d.

calebchongsj committed:
Merge branch 'branch-Level-8'

calebchongsj committed:
More-OOP increment. Refactor the code to extract out closely related code as classes.

calebchongsj committed:
Merge branch 'branch-MoreOOP'

calebchongsj committed:
A-Packages increment. Refactor all the classes into one package.

calebchongsj committed:
Merge branch 'branch-Packages'

calebchongsj committed:
Merge branch 'add-gradle-support'

calebchongsj committed:
J-Unit increment. Add JUnit tests to test the behaviour of the code

calebchongsj committed:
Merge branch 'branch-JUnit'

calebchongsj committed:
Add JavaDoc comments to the code

calebchongsj committed:
Fix the merging conflict issue

calebchongsj committed:
Merge branch 'branch-A-JavaDoc'

calebchongsj committed:
Merge branch 'branch-A-CodingStandard'

calebchongsj committed:
Level-9 increment. Add the search functionality for user keyword

calebchongsj committed:
Merge branch 'branch-Level-9'

calebchongsj committed:
Implement Java-FX in build.gradle

calebchongsj committed:
Level-10 increment. Add GUI to the chatbot

calebchongsj committed:
Merge branch 'branch-Level-10'

calebchongsj committed:
A-CheckStyle increment. Add the checkstyle plugin into Intellij IDEA

calebchongsj committed:
Merge branch 'branch-CheckStyle'

voldiemort commented on src/main/java/myapp/HelperBuddy/Deadline.java:
```diff
@@ -0,0 +1,85 @@
+import java.time.format.DateTimeParseException;
+
+/**
+ * Deadline class representing a Deadline task
```
Would be good to standardise punctuation at the end of sentences in Javadoc

voldiemort commented on src/main/java/myapp/HelperBuddy/Deadline.java:
```diff
@@ -0,0 +1,85 @@
+    private final LocalDateTime deadlineBy;
+
+    /**
+     * Constructs a {@code Deadline} task with the specified description and deadline.
```
I like how you use the @code here

voldiemort commented on src/main/java/myapp/HelperBuddy/Deadline.java:
```diff
@@ -0,0 +1,85 @@
+     * description, and optionally, the deadline.
+     * If the deadline is missing or incorrectly formatted, a warning is printed, and the deadline will be null.
+     *
+     * @param taskData The string representing the Deadline task.
```
I like how detailed the comment is. Given that the comment covered the details of the parameter quite comprehensively, maybe the @param line here could be omitted?

Same sentiment on some of the other method comments

voldiemort commented on src/main/java/myapp/HelperBuddy/DialogBox.java:
```diff
@@ -0,0 +1,84 @@
+    /**
+     * Flips the dialog box horizontally, reversing the order of the text and image.
+     * This method is used to align the bot's dialog differently from the user's dialog.
+     */
```
I like that you took the time to insert Javadoc comments for private methods as well even though it's only required for public methods and classes

voldiemort reviewed (commented):
Looks good overall.
On the whole, perhaps you might want to start classifying your classes into different packages based on their use cases.

AngPengXuan commented on src/main/java/myapp/HelperBuddy/Deadline.java:
```diff
@@ -0,0 +1,85 @@
+ * Deadline class representing a Deadline task
+ */
+public class Deadline extends Task {
+    private final LocalDateTime deadlineBy;
```
Good that you took care to ensure the deadlineBy is immutable an private.

AngPengXuan commented on src/main/java/myapp/HelperBuddy/Storage.java:
```diff
@@ -0,0 +1,66 @@
+            }
+            writer.close();
+        } catch (IOException e) {
+            System.out.println("An error occurred while saving tasks to file.");
```
Maybe print the error e here also? Otherwise it might be difficult for you to debug when things go wrong.

AngPengXuan commented on src/main/java/myapp/HelperBuddy/Task.java:
```diff
@@ -0,0 +1,66 @@
+     * Marks the task as not done.
+     * This method sets the completion status of the task to false.
+     */
+    public void markUndone() {
```
Perhaps markDone and markUndone can be combined into something like toggleMark? using
this.isDone = !this.isDone

AngPengXuan commented on src/test/java/myapp/HelperBuddy/DeadlineTest.java:
```diff
@@ -0,0 +1,43 @@
+        String[] parts = taskData.split(" \\| ");
+        String description = parts[2].trim();
+        LocalDateTime deadlineBy = null;
+        if (parts.length > 3) {
```
Not sure how necessary this if statement is, since parts is defined by you and will always be the same?

AngPengXuan commented on src/test/java/myapp/HelperBuddy/ToDoTest.java:
```diff
@@ -0,0 +1,29 @@
+        task.markDone();
+        String expected = "T | " + (task.getDone() ? "1" : "0") + " | " + task.getDescription();
+        String actual = "T | 1 | evening workout";
+        assertEquals(expected, actual);
```
There are a lot of similarities between testFormating and testParsing, maybe can combine into one function to avoid repetition?

AngPengXuan reviewed (commented):
Good work! I run your code through checkstyle and there were some wrong ordering of imports in DialogBox.java and Ui.java, but otherwise no other issues! Your code also followed the proper naming guidelines!

calebchongsj committed:
Amend checkstyle configurations for Java GUI application

calebchongsj committed:
Add assertions for validation and include JAR file for application.

The `Ui` class currently lacks assertions to validate inputs, and the necessary JAR file for the application is missing.
This situation may lead to undetected issues and improper application execution.

Add assertions throughout the classes to ensure input validation and code integrity.
Include the required JAR file in the project to support correct application functionality.

Assertions provide immediate feedback on potential issues, enhancing code reliability and correctness.
Including the JAR file guarantees that all dependencies are present for proper application execution.

Test the application to confirm that assertions are validating the code as intended
and that the JAR file integration is successful.

calebchongsj committed:
Improve code quality by reducing nesting and implementing the happy
path.

The current implementation has deep nesting and relies heavily
on global variables, which makes the code harder to read and maintain.
Additionally, error conditions are intertwined with the main execution
flow, obscuring the happy path.

Refactor the `Ui` class to:
- Use guard clauses to handle error conditions early and exit gracefully.
- Minimize the scope of variables to only where they are needed.
- Simplify the main execution path to make it more prominent and clear.

This approach clarifies the primary logic by separating error handling
and reduces implicit dependencies by limiting global variable usage.
As a result, the code becomes more readable and maintainable.

No functional changes are introduced; the focus is on
improving code quality and readability.

calebchongsj committed:
Refactor task list and search results handling with Java Streams.

The current implementation of printing the task list message and
showing search results uses manual iteration, which is less concise
and harder to read.

Refactor the `Ui` class to:
- Use Java Streams to generate the task list message,
streamlining the process and improving readability.
- Apply Java Streams to handle search results,
making the code more expressive and easier to maintain.

This change enhances code clarity and leverages modern Java features
to simplify data processing. No functional changes are introduced since
the focus is on code efficiency and readability.

calebchongsj committed:
Merge pull request #1 from calebchongsj/branch-A-Assertions

Add assertions for validation and include JAR file for application.

calebchongsj committed:
Merge branch 'branch-A-CodeQuality'

# Conflicts:
#	src/main/java/myapp/HelperBuddy/Event.java
#	src/main/java/myapp/HelperBuddy/Ui.java

calebchongsj committed:
Merge branch 'branch-A-Streams'

# Conflicts:
#	src/main/java/myapp/HelperBuddy/Ui.java

calebchongsj committed:
Amend the Event class to fix the merge conflict errors

calebchongsj committed:
Resolve merge conflicts in main branch.

The `main` branch currently contains unresolved merge conflicts
which prevent smooth integration of changes. This situation needs to
be addressed to ensure that the branch is stable and reflects
the latest codebase accurately.

Resolve all merge conflicts in the `main` branch. Ensure that the
branch incorporates changes from both conflicting branches correctly
and maintains functionality.

Resolving merge conflicts is essential for maintaining a stable
and consistent codebase, allowing for seamless integration
of updates and preventing future issues.

calebchongsj committed:
Improve code quality by nesting and happy path

The current implementation has deep nesting and relies heavily
on global variables, which makes the code harder to read and maintain.
Additionally, error conditions are intertwined with the main execution
flow, obscuring the happy path.

No functional changes are introduced since the focus is on
improving code quality.

calebchongsj committed:
Merge branch 'branch-A-CodeQuality'

# Conflicts:
#	src/main/java/myapp/HelperBuddy/Ui.java

calebchongsj committed:
Amend some code changes for the Java GUI chatbot

calebchongsj committed:
Add Continuous Integration tool for GitHub

calebchongsj committed:
Add Continuous Integration tool using GitHub

calebchongsj committed:
Merge pull request #4 from calebchongsj/branch-A-CI

branch-A-CI

calebchongsj committed:
Enhance appearance of Java GUI application

calebchongsj committed:
Merge branch 'branch-A-BetterGui'

calebchongsj committed:
Add a unique personality to my chatbot

calebchongsj committed:
Merge branch 'branch-A-Personality'

calebchongsj committed:
Update the final version of my chatbot

calebchongsj committed:
Add image of my Java GUI chatbot

calebchongsj committed:
Add Ui picture and simple user guide for QuirkBot

calebchongsj committed:
Merge branch 'branch-A-UserGuide'

calebchongsj committed:
Update new Ui image and user guide

calebchongsj committed:
Amend build.gradle to include JAR file

calebchongsj committed:
Add the JAR file for my QuirkBot GUI application

calebchongsj committed:
Merge branch 'branch-A-Release'

calebchongsj committed:
Make GitHub release for JAR file

calebchongsj committed:
Reformat and clean up JavaFX GUI code

calebchongsj committed:
Remove QuirkBot.jar file from GitHub repository
